File: spring-boot/spring-boot/src/main/java/com/example/controller/HelloService.java
package com.example.controller;

import java.time.LocalDateTime;

import org.springframework.stereotype.Component;
import org.springframework.validation.annotation.Validated;

import com.example.ExceptionHandlerImpl;

import com.linecorp.armeria.server.annotation.ExceptionHandler;
import com.linecorp.armeria.server.annotation.Get;
import com.linecorp.armeria.server.annotation.Param;
import com.linecorp.armeria.server.annotation.Post;
import com.linecorp.armeria.server.annotation.ProducesJson;

import jakarta.validation.Valid;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotNull;

@Component
@Validated
@ExceptionHandler(ExceptionHandlerImpl.class)
public class HelloService {
    @Get("/hello1")
    public String hello1(@Param String name) {
        return "Hello, %s!".formatted(name);
    }

    @Get("/hello2")
    public String hello2(@Param @Min(1) int number) {
        return "Hello, %d!".formatted(number);
    }

    @Post("/hello3")
    @ProducesJson
    public HelloResponse hello3(@Valid HelloRequest request) {
        return new HelloResponse(request.message, LocalDateTime.now());
    }

    public record HelloRequest(@NotNull String message) {}

    public record HelloResponse(String message, LocalDateTime dateTime) {}
}

File: spring-armeria-kotlin/src/main/kotlin/com/example/HelloService.kt
package com.example;


import com.linecorp.armeria.server.annotation.Default
import com.linecorp.armeria.server.annotation.Get
import com.linecorp.armeria.server.annotation.Param
import com.linecorp.armeria.server.annotation.ProducesJson
import kotlinx.coroutines.delay
import org.springframework.stereotype.Component

@Component
class HelloService {

    @Get("/hello1")
    fun hello1(): String = "Hello!"

    @Get("/hello2")
    @ProducesJson
    suspend fun hello2(@Param @Default("hirakida") name: String): Response =
        delayedHello(name)

    private suspend fun delayedHello(name: String): Response {
        delay(1000)
        return Response("Hello, $name!")
    }

    data class Response(val message: String)
}
